### PR TITLE
Operations should provide a RawResposne

### LINKED ISSUE #1739: Operation<T> will have a `GetRawResponse()` before we GA,
Operation<T> will have a `GetRawResponse()` before we GA,
and calling `Poll()` will return the status instead of the `rawResponse`

_Originally posted by @vhvb1989 in https://github.com/Azure/azure-sdk-for-cpp/pull/1736#discussion_r582671237_

### PR DESCRIPTION
**Pull request by RickWinter**
Fixes #1739 
Added the pure virtual method
On storage operation types they currently return null.  Storage will need to update their implementation to return the actual rawResponse.

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [x] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

File: CMakeSettings.json [hunk 1]
```diff
@@ -218,11 +203,6 @@
           "name": "BUILD_PERFORMANCE_TESTS",
           "value": "True",
           "type": "BOOL"
-        },
-        {
-          "name": "CMAKE_BUILD_TYPE",
-          "value": "MinSizeRel",
-          "type": "STRING"
```
Comment (ahsonkhan) [inline CMakeSettings.json:225]: I was seeing similar warnings locally about `CMAKE_BUILD_TYPE` being defined twice. Do you know where this is already defined?

File: sdk/core/azure-core/inc/azure/core/operation.hpp [hunk 2]
```diff
@@ -51,6 +51,8 @@ namespace Azure { namespace Core {
      */
     virtual std::string GetResumeToken() const = 0;
 
+    virtual Azure::Core::Http::RawResponse* GetRawResponse() const = 0;
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/operation.hpp:54]: Need doc comments.

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:54]: Should this be as follows to match `Response<T>`?
https://github.com/Azure/azure-sdk-for-cpp/blob/3b88c7bb83b04382f5eb1fbf62a1bbecd6ec3620/sdk/core/azure-core/inc/azure/core/response.hpp#L48-L53

```suggestion
    virtual Azure::Core::Http::RawResponse& GetRawResponse() const = 0;
```

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:54]: Add  a changelog entry.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp [hunk 3]
```diff
@@ -77,6 +77,13 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
         Azure::Core::Response<Azure::Security::KeyVault::Keys::DeletedKey> response);
 
   public:
+    /**
+     * @brief Get the #Azure::Core::Http::RawResponse of the operation request.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
+     * exist.
+     */
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:84]: Add this doc comment to the base type, in operation.hpp.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp [hunk 4]
```diff
@@ -77,6 +77,13 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
         Azure::Core::Response<Azure::Security::KeyVault::Keys::DeletedKey> response);
 
   public:
+    /**
+     * @brief Get the #Azure::Core::Http::RawResponse of the operation request.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:82]: When would it be null? When wouldn't  a response not exist?

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp [hunk 5]
```diff
@@ -81,6 +81,16 @@ namespace Azure { namespace Storage { namespace Blobs {
 
   public:
     Models::GetBlobPropertiesResult Value() const override { return m_pollResult; }
+    /**
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp:84]: nit: add a new line in between the previous method and the comment.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp [hunk 6]
```diff
@@ -81,6 +81,16 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+     * @brief Get the #Azure::Core::Http::RawResponse of the operation request.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
+     * exist.
+     */
+    Azure::Core::Http::RawResponse* GetRawResponse() const override
+    {
+      // TODO: Fix to return the rawResponse
+      return nullptr;
+    }
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp:93]: FYI @Jinming-Hu

File: CMakeSettings.json (hunk 1 above)
Comment (RickWinter) [reply CMakeSettings.json:225]: @vhvb1989 who added them.  I removed cause they are noise.

File: sdk/core/azure-core/inc/azure/core/operation.hpp (hunk 2 above)
Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:54]: Will do.  Though customers should NEVER use the type.  It should be moved to internal as well.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 4 above)
Comment (RickWinter) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:82]: If you've extracted it already.  We don't allow Extracting it at the moment but we may want to support taking ownership of the contents at some point.

**Comment by ahsonkhan** [comment]
Looks like some networking/infra issue causing the test failures:
https://dev.azure.com/azure-sdk/public/_build/results?buildId=769740&view=logs&j=5f43df56-f144-584b-8cac-2916b6897a0a&t=a63c70f6-59e4-531a-efba-82f28f72650d&l=3392

```text
164: unknown file: error: C++ exception with description "Fail to get a new connection for: anglesharp.azurewebsites.net. Couldn't resolve host name" thrown in the test body.
164: [  FAILED  ] Test/TransportAdapter.getChunkWithStream/libCurl, where GetParam() = 
```
```text
2021-03-06T02:21:05.8332379Z 155: [ RUN      ] Test/TransportAdapter.headWithStream/winHttp
2021-03-06T02:21:05.9750114Z 155: unknown file: error: C++ exception with description "Error while sending a request. Error Code: 12007." thrown in the test body.
```
```text
2021-03-06T02:21:06.1506806Z 156: unknown file: error: C++ exception with description "Fail to get a new connection for: httpbin.org. Couldn't resolve host name" thrown in the test body.
2021-03-06T02:21:06.1509619Z 156: [  FAILED  ] Test/TransportAdapter.headWithStream/libCurl, where GetParam()
```

FYI @vhvb1989, @weshaggard - some brittleness in our tests relying on things like httpbin.

```text

The following tests FAILED:
	155 - azure-core.Test/TransportAdapter.headWithStream/winHttp (Failed)
	156 - azure-core.Test/TransportAdapter.headWithStream/libCurl (Failed)
	157 - azure-core.Test/TransportAdapter.putWithStream/winHttp (Failed)
	158 - azure-core.Test/TransportAdapter.putWithStream/libCurl (Failed)
	159 - azure-core.Test/TransportAdapter.deleteRequestWithStream/winHttp (Failed)
	160 - azure-core.Test/TransportAdapter.deleteRequestWithStream/libCurl (Failed)
	161 - azure-core.Test/TransportAdapter.patchWithStream/winHttp (Failed)
	162 - azure-core.Test/TransportAdapter.patchWithStream/libCurl (Failed)
	163 - azure-core.Test/TransportAdapter.getChunkWithStream/winHttp (Failed)
	164 - azure-core.Test/TransportAdapter.getChunkWithStream/libCurl (Failed)
	165 - azure-core.Test/TransportAdapter.createResponseT/winHttp (Failed)

```

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 4 above)
Comment (ahsonkhan) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:82]: Wouldn't we abort/throw at the point because that would be a dev-time bug in the calling code? It is odd that we'd be returning null raw responses in what I consider a buggy usage.

File: sdk/core/azure-core/inc/azure/core/operation.hpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:54]: `Operation<T>`? Really? That's surprising then that .NET put it at the root namespace:
https://github.com/Azure/azure-sdk-for-net/blob/3e8a709af4c78c856cc753a8ed5a48c95a1fa92a/sdk/core/Azure.Core/src/OperationOfT.cs#L11-L21

Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:54]: Yea, C++ has value semantics and that means that if a user can create an operation<T> they could then replace the guts of the derived type.  Remember that in .NET<T> its a generic type which is not the same as a template.

**Review by Jinming-Hu — APPROVED** [review]

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 4 above)
Comment (RickWinter) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:82]: We could throw if that's the design we want.  We want to dive deeper in the design.  As the code is written its correct.  BTW I filed a bug for our usage pattern on HttpTransports which has undefined behavior

File: sdk/core/azure-core/CHANGELOG.md [hunk 7]
```diff
@@ -7,6 +7,7 @@
 - Removed `Azure::Core::Http::HttpPipeline` by making it internal, used only within the SDK.
 - Split `Azure::Core::RequestConditions` into `Azure::Core::MatchConditions` and `Azure::Core::ModifiedConditions`.
 - Removed `TransportKind` enum from `Azure::Core::Http`.
+- Added `Azure::Core::Operation<T>::GetRawResponse`
```
Comment (ahsonkhan) [inline sdk/core/azure-core/CHANGELOG.md:10]:
```suggestion
- Added `Azure::Core::Operation<T>::GetRawResponse().`
```

File: sdk/core/azure-core/inc/azure/core/operation.hpp [hunk 8]
```diff
@@ -51,6 +51,14 @@ namespace Azure { namespace Core {
      */
     virtual std::string GetResumeToken() const = 0;
 
+    /**
+     * @brief Get the raw HTTP response.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
+     * exist.
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/operation.hpp:57]: Since we haven't designed for this yet (and because I don't think we should return null anyway), let's not mention that in the docs **yet**.

```suggestion
     * @return A pointer to #Azure::Core::Http::RawResponse.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp [hunk 9]
```diff
@@ -82,6 +82,18 @@ namespace Azure { namespace Storage { namespace Blobs {
   public:
     Models::GetBlobPropertiesResult Value() const override { return m_pollResult; }
 
+    /**
+     * @brief Get the raw HTTP response.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
+     * exist.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp:88]:
```suggestion
     * @return A pointer to #Azure::Core::Http::RawResponse.
```

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp [hunk 10]
```diff
@@ -198,6 +198,18 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
   public:
     Models::GetShareFilePropertiesResult Value() const override { return m_pollResult; }
 
+    /**
+     * @brief Get the raw HTTP response.
+     * @return A pointer to #Azure::Core::Http::RawResponse or null if the RawResponse does not
+     * exist.
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp:204]:
```suggestion
     * @return A pointer to #Azure::Core::Http::RawResponse or null.
```

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp [hunk 11]
```diff
@@ -198,6 +198,18 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
+     * @note Does not give up ownership of the RawResponse
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp:205]:
```suggestion
     * @note Does not give up ownership of the RawResponse.
```

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp [hunk 12]
```diff
@@ -82,6 +82,18 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+     * @note Does not give up ownership of the RawResponse
```
Comment (ahsonkhan) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp:89]:
```suggestion
     * @note Does not give up ownership of the RawResponse.
```

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 3 above)
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:84]: Add `@note Does not give up ownership of the RawResponse.`

File: sdk/core/azure-core/inc/azure/core/operation.hpp [hunk 13]
```diff
@@ -51,6 +51,14 @@ namespace Azure { namespace Core {
...
+     * @note Does not give up ownership of the RawResponse
```
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/operation.hpp:58]:
```suggestion
     * @note Does not give up ownership of the RawResponse.
```

File: sdk/core/azure-core/inc/azure/core/operation.hpp (hunk 2 above)
Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:60]: Are we going to return a pointer or a reference? `Response<T>` returns a reference.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp [hunk 14]
```diff
@@ -77,6 +77,13 @@ namespace Azure { namespace Security { namespace KeyVault { namespace Keys {
...
+     * exist.
```
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:83]:
```suggestion
     * @return A pointer to #Azure::Core::Http::RawResponse or null.
```

**Review by ahsonkhan — APPROVED** [review]
Left some suggestions to fix up doc comments.

Otherwise, looks good.

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 4 above)
Comment (ahsonkhan) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:82]: Ah I see, just saw the issue: https://github.com/Azure/azure-sdk-for-cpp/issues/1805

File: sdk/core/azure-core/inc/azure/core/operation.hpp (hunk 2 above)
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:60]: +1 for reference and maybe we should have two overloads

```
virtual Azure::Core::Http::RawResponse const& GetRawResponse() const = 0;
virtual Azure::Core::Http::RawResponse& GetRawResponse() = 0;
```

Comment (RickWinter) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:59]: If we do a reference return then we need to throw if the ptr is null.  Otherwise its undefined behavior.  

What do we do if a user extracts the m_rawResponse and then calls GetRawResponse?  Users need to try/catch all calls to GetRawResponse because they may or may not have access to the rawresponse.

Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:59]: > If we do a reference return then we need to throw if the ptr is null. Otherwise its undefined behavior.
> 
> What do we do if a user extracts the m_rawResponse and then calls GetRawResponse? Users need to try/catch all calls to GetRawResponse because they may or may not have access to the rawresponse.

same question for `Response<T>`

File: CMakeSettings.json [hunk 15]
```diff
@@ -137,11 +132,6 @@
...
```
Comment (vhvb1989) [inline CMakeSettings.json:144]: Why did you remove this configuration from the Perf profiles?  This is the Release config for Perf.

File: CMakeSettings.json (hunk 1 above)
Comment (vhvb1989) [reply CMakeSettings.json:225]: lol, after removing it, the Perf profiles for VS won't use use the most optimized build. 
As long as you don't use this profiles, how is it noise?

File: sdk/core/azure-core/inc/azure/core/operation.hpp (hunk 2 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:59]: Well, we solved this on Nullable<T> by aborting if T is empty and we call Get().
Probably we should do the same with Response<T> and Operation<T>

Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/operation.hpp:59]: At least this should return a pointer to a constant RawResponse?  or do we want to allow to update the RawResponse?

File: sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp (hunk 3 above)
Comment (vhvb1989) [reply sdk/keyvault/azure-security-keyvault-keys/inc/azure/keyvault/keys/delete_key_operation.hpp:83]: I can call this method, and update the RawResponse :).  It might mess up the internal state of the Operation<T>.
If we don't want to return a `const&` or a `RawResponse const*`, then we should return a a copy of the entire RawResponse.

File: sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp [hunk 16]
```diff
@@ -82,6 +82,17 @@ namespace Azure { namespace Storage { namespace Blobs {
...
     Models::GetBlobPropertiesResult Value() const override { return m_pollResult; }
 
+    /**
+     * @brief Get the raw HTTP response.
+     * @return A pointer to #Azure::Core::Http::RawResponse.
+     * @note Does not give up ownership of the RawResponse.
+     */
+    Azure::Core::Http::RawResponse* GetRawResponse() const override
+    {
+      // TODO: Fix to return the rawResponse
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/blob_responses.hpp:92]: You have just violated your own rule about never merging a `TODO:` comment.  =\

File: sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp [hunk 17]
```diff
@@ -198,6 +198,17 @@ namespace Azure { namespace Storage { namespace Files { namespace Shares {
...
 
+    /**
+     * @brief Get the raw HTTP response.
+     * @return A pointer to #Azure::Core::Http::RawResponse.
+     * @note Does not give up ownership of the RawResponse.
+     */
+    Azure::Core::Http::RawResponse* GetRawResponse() const override
+    {
+      // TODO: Fix to return the rawResponse
+      return nullptr;
```
Comment (vhvb1989) [inline sdk/storage/azure-storage-files-shares/inc/azure/storage/files/shares/share_responses.hpp:209]: image
